### PR TITLE
transaction_log Metricset for MSSQL Module (also db metricset removal)

### PR DESCRIPTION
**Pull request by sayden**
## Summary
* `transaction_log` Metricset in MSSQL module. 
* Replaces / Removes "old" `db` Metricset.
* Removes references to "old" `fetcher.go` file with the abstractions that we'll use in the SQL helper.

### `transaction_log` Metricset
Includes information about the log space usage and general log stats like backup size, time elapsed since last backup or log size since last backup.

### Removal / Renaming of `db` Metricset
After careful anaysis, `db` naming had 3 problems:
* You can merely put everything in a `db` metricset because, all in all, 90% of metrics are part of a database.
* `db` means nothing and everything. A user will face the question "do I want to activate `db` metricset? Yes, of course, always! I want metrics from the database!" So the answer was always "Yes".
* 80% of the data that was included was about the database log and file space usage so I left only the metrics that were about the log (and left file space usage for a different metricset).

### Removal of `fetcher.go` file
I was using an abstraction layer to perform, parse and emit metrics based on the queries that were provided. We decided to make it more explicit and less abstract to improve clarity and maintenance. The only metricset using the abstraction layer was `db` so now that it's removed, the file could be removed too.

It also forced to remove a forgotten reference in `performance` metricset which is included in this PR.

## GA

* [X] Supported versions are documented
* [X] Supported operating systems are documented (if applicable)
* [X] Integration tests
* [X] System tests
* [X] Automated checks that all fields are documented
* [X] Documentation
* [X] Example `data.json` exists and an automated way to generate it exists (`go test -data`)
* [X] Fields follow naming conventions: https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html
* [ ] Dashboards exists (if applicable)
* [ ] Kibana Home Tutorial (if applicable)

### CODE REVIEW

File: x-pack/metricbeat/module/mssql/connection.go [hunk 1]
```diff
@@ -7,6 +7,9 @@ package mssql
 import (
 	"database/sql"
 
+	// Register driver.
+	_ "github.com/denisenkom/go-mssqldb"
```
Comment (sayden) [inline x-pack/metricbeat/module/mssql/connection.go]: The driver was previously registered in `fetcher.go` file, which is now removed.

File: x-pack/metricbeat/module/mssql/log/data.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+package db
+
+import (
+	"time"
+
+	s "github.com/elastic/beats/libbeat/common/schema"
+	c "github.com/elastic/beats/libbeat/common/schema/mapstrstr"
+)
+
+var (
```
Comment (sayden) [inline x-pack/metricbeat/module/mssql/log/data.go]: File included for clarity but it's not being used.

File: x-pack/metricbeat/module/mssql/performance/performance.go [hunk 3]
```diff
@@ -42,9 +42,8 @@ func init() {
 // interface methods except for Fetch.
 type MetricSet struct {
 	mb.BaseMetricSet
-	log     *logp.Logger
-	fetcher *mssql.Fetcher
```
Comment (sayden) [inline x-pack/metricbeat/module/mssql/performance/performance.go]: This was the reference that needed to be removed after deleting `fetcher.go`

File: x-pack/metricbeat/module/mssql/_meta/docs.asciidoc [hunk 4]
```diff
@@ -8,8 +8,12 @@ The module is being tested with https://hub.docker.com/r/microsoft/mssql-server-
 [float]
 == Metricsets
 
-The following Metricset is already included:
+The following Metricsets are already included:
 
-=== `db`
+=== `log`
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/_meta/docs.asciidoc]: Sounds like next we need the MSSQL module for the actual logs :-)

File: x-pack/metricbeat/module/mssql/log/_meta/data.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                "used": {
+                    "bytes": 688128,
+                    "pct": 32.94117736816406
+                }
+            },
+            "stats": {
+                "active_size": {
+                    "bytes": 94207.213568
+                },
+                "backup_time": "1900-01-01T00:00:00Z",
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/log/_meta/data.json]: Interesting timestamp :-)

File: x-pack/metricbeat/module/mssql/log/log.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+	if err != nil {
+		return
+	}
+	defer closeRows(rows)
+
+	for rows.Next() {
+		var res logSpace
+		if err = rows.Scan(&res.id, &res.totalLogSizeInBytes, &res.usedLogSpaceInBytes, &res.usedLogSpaceInPercent,
+			&res.logSpaceInBytesSinceLastBackup); err != nil {
+			return
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/log/log.go]: Should we continue or return here? Do we expect the error to apply all scans?

File: x-pack/metricbeat/module/mssql/log/log.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+			},
+			"used": common.MapStr{
+				"bytes": res.usedLogSpaceInBytes,
+				"pct":   res.usedLogSpaceInPercent,
+			},
+			"since_last_backup": common.MapStr{
+				"bytes": res.logSpaceInBytesSinceLastBackup,
+			},
+		}
+		// According to docs "All transaction log files are combined." Means,
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/log/log.go]: What is this related to?

File: x-pack/metricbeat/module/mssql/log/log.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+			},
+			"recovery_size": common.MapStr{
+				"bytes": res.recoverySizeMB,
+			},
+		}
+
+		if res.sinceLastBackupMB != nil {
+			result.Put("since_last_backup.bytes", *res.sinceLastBackupMB*1048576)
+		}
+		// According to docs "All transaction log files are combined." Means,
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/log/log.go]: ?

File: x-pack/metricbeat/module/mssql/log/log.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+		spaceUsage, err := m.getLogSpaceUsageForDb(db.name)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["space_usage"] = spaceUsage
+		}
+
+		stats, err := m.getLogStats(db)
+		if err != nil {
+			reporter.Error(err)
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/log/log.go]: Also, if some of these errors happen, do we want to continue?

File: x-pack/metricbeat/module/mssql/log/_meta/data.json (hunk 5 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/_meta/data.json]: In fact, it's correct! It seems that it's the default value that MSSQL returns when no value is set (and it's not nil cell).

image

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 6 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: This query only returns a single result (according to docs) so an error in the first scan is a "no data returned".

If I don't return here, it might look like I have overlook at this. My future me (or any of us) might think that the error is not properly handled.

I can also report error, log it or continue (and it will return with no data nor result, which is the same than now).

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 7 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Same here. A message for the "future me" (or any of us). By just looking at the code it might look that the result is being overriden in every scan but, in fact, it just returns a single result.

Suggestions welcome :)

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 6 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Suggestions welcome :)

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 8 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Same as above. Suggestions welcome :)

File: CHANGELOG.next.asciidoc [hunk 10]
```diff
@@ -140,6 +140,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Add DeDot for kubernetes labels and annotations. {issue}9860[9860] {pull}9939[9939]
 - Add more meaningful metrics to 'performance' Metricset on 'MSSQL' module {pull}10011[10011]
 - Rename some fields in `performance` Metricset on MSSQL module to match the updated documentation from Microsoft {pull}10074[10074]
+- Rename 'db' Metricset to 'log' in MSSQL Metricbeat module {pull}10109[10109]
```
[the remark this answers is not in the document]
Comment (sayden) [reply CHANGELOG.next.asciidoc]: I have actually been thinking about this. This were my thoughts:
- `transaction_log`: It might confuse user thinking that it's some metric related to transactions but it's not. It's the internal log of data and operations that must be processed and flushed to _data_ files. Here's a nice explanation https://www.c-sharpcorner.com/article/learn-about-sql-server-architecture/
- `data_log`: Means nothing and everything
- `temporal_log`: I thought "maybe" but it's actually giving lot of meaning to something that maybe it's not correctly described as "temporal".
- `log`: Well... they are not "Da Logs" in the Elasticsearch term but it should not be not possible to confuse

File: x-pack/metricbeat/module/mssql/_meta/docs.asciidoc [hunk 11]
```diff
@@ -8,8 +8,12 @@ The module is being tested with https://hub.docker.com/r/microsoft/mssql-server-
...
 
-`db` Metricset fetches a set of values from https://docs.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/sys-dm-db-log-space-usage-transact-sql?view=sql-server-2017[log space usage] which returns space usage information for the transaction log for each different database.
+`log` Metricset fetches information about the logs of each MSSQL database in the monitored instance. All data is extracted from the https://docs.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/database-related-dynamic-management-views-transact-sql?view=sql-server-2017[Database Dynamic Management Views]
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/docs.asciidoc]: Yeah, re-reading now after a day it looks confusing.

File: x-pack/metricbeat/module/mssql/_meta/fields.yml [hunk 12]
```diff
@@ -7,3 +7,13 @@
...
+        - name: database
+          type: group
+          description: The database that the metrics is being referred to
+          fields:
+            - name: id
+              type: long
+              description: Unique ID of the database inside MSSQL
+            - name: name
+              type: keyword
+              description: Name of the database
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/fields.yml]: :wink:

File: x-pack/metricbeat/module/mssql/log/log.go [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+
+func (m *MetricSet) getLogSpaceUsageForDb(dbName string) (result common.MapStr, err error) {
+
+	rows, err := m.db.Query(fmt.Sprintf(`USE %s; SELECT * FROM sys.dm_db_log_space_usage;`, dbName))
+	if err != nil {
+		return
+	}
+	defer closeRows(rows)
+
+	for rows.Next() {
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Uhmmm... not a bad idea, but I'll still have to leave a message for our future us so that they don't think that it's a mistake :smile:

File: x-pack/metricbeat/module/mssql/log/log.go [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+// Close the connection to the server at the engine level
+func (m *MetricSet) Close() error {
+	return m.db.Close()
+}
+
+func (m *MetricSet) getLogSpaceUsageForDb(dbName string) (result common.MapStr, err error) {
+
+	rows, err := m.db.Query(fmt.Sprintf(`USE %s; SELECT * FROM sys.dm_db_log_space_usage;`, dbName))
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: `?` placeholders doesn't work well with this driver, I don't know why. I couldn't find a way to make it work but using `Sprintf`

File: x-pack/metricbeat/module/mssql/log/data.go (hunk 2 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/data.go]: If for any reason we want to know to aspect of the data that is being sent to ES, we tend to go to look the `data.go` file. Of course, we can also look at the `data.json` generated file but there are not guarantees that it's updated (neither this file).

In short, it's to help any maintainer that it's not me :)

File: x-pack/metricbeat/module/mssql/log/log.go [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+			},
+		}
+		// According to docs "All transaction log files are combined." Means,
+		// a single result is returned
+	}
+
+	return
+}
+func (m *MetricSet) getLogStats(db dbInfo) (result common.MapStr, err error) {
+
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Same answer as above, I'm fine with changing but let's confirm that it's what we want, I'll wait your answer :smile:

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Ok, you are right, I'll add some context. About the second question: Yes, I'd say that we want to continue. It's very unlikely that one of the objects is going to fail. Almost impossible that two of them will fail and it's not because of a network error (that should _explode_ somewhere else).

What I can do here is to handle the situation where `space_usage` and `stats` aren't filled to not try to report an empty event.

File: CHANGELOG.next.asciidoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (sayden) [reply CHANGELOG.next.asciidoc]: I have just discover that Microsoft also call it "transaction log" even when it's the log for everything.

https://docs.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/sys-dm-db-log-space-usage-transact-sql?view=sql-server-2017

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 7 above)
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/log/log.go]: I think I get what you commented here more then what is in the comment itself. Also the location is a bit odd. I normally expect the comment before the code line that it explains.

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 6 above)
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/log/log.go]: Does this error have a specific type we can detect?

Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: No. It seems that they are plain text wrapped with more error handling from the Go SQL library itself:

- Error wrapping in Go's SQL: https://golang.org/src/database/sql/sql.go?s=73165:73887#L2914
- Some error returned by the MSSQL library itself: https://github.com/denisenkom/go-mssqldb/blob/master/uniqueidentifier.go#L37

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 8 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Following the converstation above, I'll later fix here too.

File: CHANGELOG.next.asciidoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (sayden) [reply CHANGELOG.next.asciidoc]: @jsoriano maybe you can check the names mentioned above? @ruflin do you have any idea too?

File: x-pack/metricbeat/module/mssql/_meta/docs.asciidoc (hunk 11 above)
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/docs.asciidoc]: Rephrased to _`log` Metricset fetches information about the operation and transaction log of each MSSQL_ database

File: x-pack/metricbeat/module/mssql/log/_meta/docs.asciidoc [hunk 16]
```diff
@@ -0,0 +1,11 @@
+`log` Metricset fetches information about the logs of each MSSQL database in the monitored instance. All data is extracted from the https://docs.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/database-related-dynamic-management-views-transact-sql?view=sql-server-2017[Database Dynamic Management Views]
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/_meta/docs.asciidoc]: Written as above: _about the operation and transaction log_

File: CHANGELOG.next.asciidoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply CHANGELOG.next.asciidoc]: I'd go for `transactionlog` (without underscore), specially if Microsoft also calls it this way. They also mention that this is about transactions: https://docs.microsoft.com/en-us/sql/relational-databases/logs/the-transaction-log-sql-server?view=sql-server-2017

> this is not Filebeat but Metricbeat.

The separation between them can be removed in the future.

File: x-pack/metricbeat/module/mssql/_meta/docs.asciidoc (hunk 11 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/_meta/docs.asciidoc]: Why operation log? The log seems to be called _transaction log_.

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/log/log.go]: Well, the comment maybe needed in any case because I already though that there were some mistake here :smiley: And also Nicolas with the return. I think that without a loop it'd be more clear that only a row is expected.

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/log/log.go]: Ok, sounds good.

File: x-pack/metricbeat/module/mssql/log/data.go (hunk 2 above)
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/log/data.go]: `data.go` uses to contain the code for mapping the data to events, if there is no clear separation for that in this metricset I'd prefer to remove this file than to keep it with unused code.

If it is to know more about the generated data, as you mention `data.json` can be used.

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 7 above)
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/log/log.go]: No. I would expect the comment on line 134 to be something like:

```
This query only returns a single result so an error in the first scan is a "data returned".
```

File: CHANGELOG.next.asciidoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply CHANGELOG.next.asciidoc]: +1. With and without _ works for me.

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/log/log.go]: ++

File: x-pack/metricbeat/module/mssql/transaction_log/data_integration_test.go [hunk 17]
```diff
@@ -2,7 +2,7 @@
 // or more contributor license agreements. Licensed under the Elastic License;
 // you may not use this file except in compliance with the Elastic License.
 
-package db
+package transaction_log
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/transaction_log/data_integration_test.go]: I know we shouldn't use underscore but it seems accepted according to ES modules: https://github.com/elastic/beats/blob/master/metricbeat/module/elasticsearch/index_recovery/data_xpack.go#L18

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go [hunk 18]
```diff
@@ -0,0 +1,223 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package transaction_log
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: I know we shouldn't use underscore but it seems accepted according to ES modules: https://github.com/elastic/beats/blob/master/metricbeat/module/elasticsearch/index_recovery/data_xpack.go#L18

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log_integration_test.go [hunk 19]
```diff
@@ -4,7 +4,7 @@
 
 // +build integration
 
-package db
+package transaction_log
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/transaction_log/transaction_log_integration_test.go]: I know we shouldn't use underscore but it seems accepted according to ES modules: https://github.com/elastic/beats/blob/master/metricbeat/module/elasticsearch/index_recovery/data_xpack.go#L18

File: CHANGELOG.next.asciidoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (sayden) [reply CHANGELOG.next.asciidoc]: Change to `transaction_log`. I have used underscore instead of CamelCase because this way field names on JSON also follows naming conventions instead of having `transactionLog` and `transaction_log` mixed in the module.

File: x-pack/metricbeat/module/mssql/_meta/docs.asciidoc (hunk 4 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/docs.asciidoc]: Renamed to `transaction_log`

File: CHANGELOG.next.asciidoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply CHANGELOG.next.asciidoc]: Well, the alternative was not camel case but `transactionlog`. But it's true that we already have other metricsets with underscore, so as you prefer, both `transactionlog` and `transaction_log` work for me.

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: Yeah, you can ignore this.

File: CHANGELOG.next.asciidoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply CHANGELOG.next.asciidoc]: as windows uses 2 words, I would go with _

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+
+import (
+	"database/sql"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: Nit: remove newline.

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+			},
+			"recovery_size": common.MapStr{
+				"bytes": res.recoverySizeMB,
+			},
+		}
+
+		if res.sinceLastBackupMB != nil {
+			result.Put("since_last_backup.bytes", *res.sinceLastBackupMB*1048576)
+		}
+		// According to docs "All transaction log files are combined." Means,
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: This comment should be on line 165, perhaps not need if only 1 row is fetched. See other comment for content.

File: x-pack/metricbeat/module/mssql/log/data.go (hunk 2 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/data.go]: File has been removed :+1:

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 6 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: I have wrapped the error returned with some description.

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 7 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Now I'm using `QueryRow` but I have still placed some comment to clarify why

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Thanks for the suggestion @jsoriano !

File: x-pack/metricbeat/module/mssql/log/log.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/log/log.go]: Checked the content of the event to not send it empty.

File: CHANGELOG.next.asciidoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (sayden) [reply CHANGELOG.next.asciidoc]: Updated!

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go (hunk 20 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: Removed

File: x-pack/metricbeat/module/mssql/_meta/docs.asciidoc [hunk 22]
```diff
@@ -8,8 +8,12 @@ The module is being tested with https://hub.docker.com/r/microsoft/mssql-server-
 [float]
 == Metricsets
 
-The following Metricset is already included:
+The following Metricsets are already included:
 
-=== `db`
+=== `transactionLog`
```
[the remark this answers is not in the document]
Comment (sayden) [reply x-pack/metricbeat/module/mssql/_meta/docs.asciidoc]: Updated both

File: x-pack/metricbeat/module/mssql/transaction_log/_meta/data.json [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+                "used": {
+                    "bytes": 622592,
+                    "pct": 29.80392074584961
+                }
+            },
+            "stats": {
+                "active_size": {
+                    "bytes": 135167.737856
+                },
+                "backup_time": "1900-01-01T00:00:00Z",
```
Comment (jsoriano) [inline x-pack/metricbeat/module/mssql/transaction_log/_meta/data.json]: I wonder if we could leave it empty if it has this zero-like value. If someone tries to use it in dashboards or queries they will have to clean the data, we could already do it here.

File: x-pack/metricbeat/module/mssql/transaction_log/_meta/fields.yml [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+              type: long
+        - name: used
+          type: group
+          description: The occupied size of the log
+          fields:
+            - name: bytes
+              description: The occupied size of the log in bytes
+              type: long
+            - name: pct
+              description: A percentaje of the occupied size of the log as a percent of the total log size
```
Comment (jsoriano) [inline x-pack/metricbeat/module/mssql/transaction_log/_meta/fields.yml]:
```suggestion
              description: A percentaje of the occupied size of the log as a percent of the total log size
```

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go [hunk 25]
```diff
@@ -0,0 +1,217 @@
...
+
+	}
+}
+
+// Close the connection to the server at the engine level
+func (m *MetricSet) Close() error {
+	return m.db.Close()
+}
+
+func (m *MetricSet) getLogSpaceUsageForDb(dbName string) (result common.MapStr, err error) {
```
Comment (jsoriano) [inline x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: Not sure of the benefits of using named return values here, but ok.

Comment (ruflin) [reply x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: I personally prefer not to use named returns and only if it is really need like in defer functions. It makes from my perspective code more readable. @sayden could you update this?

File: x-pack/metricbeat/module/mssql/transaction_log/_meta/data.json (hunk 23 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/transaction_log/_meta/data.json]: That was the exact value returned by the database without parsing. I mean, it's the value that an experienced MSSQL user will expect to receive because it's the exact value that you get when you run the query with `sqlcmd` or with any database explorer. 

Let's try to leave our personal preferences or modifications aside because this kind of changes could confuse the users.

Please, this have been like this for days if not weeks, next time, try to make this comments earlier.

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go (hunk 25 above)
Comment (sayden) [reply x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: Changed. I see the benefits and I also have my own preferences. Now it has more code and it's not super clear that it's returning a nil at the end (if you don't look up at the return values... but I thought that this is what we wanted to avoid).

Please, this have been like this for days if not weeks, next time, try to make this comments earlier.

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+	for rows.Next() {
+		var row dbInfo
+		if err = rows.Scan(&row.name, &row.id); err != nil {
+			return
+		}
+
+		res = append(res, row)
+	}
+
+	return
```
Comment (ruflin) [inline x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: named returns?

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go (hunk 25 above)
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: Thanks for adjusting it, appreciate it. Not sure i can follow your comment above as now if I look at the end of the method it's very clear, no error is returned.

Sorry that we missed that in previous reviews but I think initially we focused on the high level details like naming before going into the details.

File: x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go (hunk 26 above)
Comment (ruflin) [reply x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: we can also delay this, no needed in this PR.

Comment (sayden) [reply x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go]: All of them!?!? Oookkk :roll_eyes: Done

**Comment by sayden** [comment]
I was running `mage fmt update` instead of `make update` from OSS folder. Does it make sense that I need to do something in the OSS folder to make an X-Pack module work? Just curious

**Comment by ruflin** [comment]
Good question. As all the docs are in OSS the answer is probably yes for now but agree it's not optimal. We should clean this up more as soon as we migrated all to mage.

@andrewkroh This might also interest you.
